### ISSUE TITLE
"AssertionError: Namespace ... does not exist" after rebuild

### ISSUE THREAD

**Issue by vitorqb** [thheller/shadow-cljs#740]
Hello! 

I'm seeing an error on a workflow I'm very used to do. The steps to reproduce would be:

1. Watch two builds (app and test)
```
npx shadow-cljs watch app test
```

2. Connect to the clj repl and drop to a cljs repl, openning the browser to create the runtime environment.
```
shadow.user> (shadow/nrepl-select :app)
To quit, type: :cljs/quit
[:selected :app]
```

3. Run some tests. To do that, I have this in my `cljs.user`:
```
(ns cljs.user
  (:require [cljs.test :refer-macros [run-tests]]
            test-ns-requires))
```

Where `test-ns-requires` is a namespace that `requires` all tests namespaces:

```
(ns test-ns-requires
  (:require
   [bar-test]
   [foo-test]
   '...)
```

So I just run:

```
(run-tests 'foo-test)
```

4. Now go to the test file and edit something, triggering a rebuild.

5. Try to test again with the same command `(run-tests 'foo-test)`. I'm getting this:

```
------ REPL Error while processing ---------------------------------------------
(run-tests 'foo-test)
Encountered error when macroexpanding cljs.test/test-ns-block.
AssertionError: Assert failed: Namespace foo-test does not exist
(ana-api/find-ns ns)
```

Here is a sequence with the repl commands and logs (with the real namespaces names used):

```
shadow.user> (shadow/nrepl-select :app)
To quit, type: :cljs/quit
[:selected :app]
cljs.user> (run-tests 'ohmycards.web.core-test)

Testing ohmycards.web.core-test

Ran 1 tests containing 2 assertions.
0 failures, 0 errors.
nil
cljs.user> (run-tests 'ohmycards.web.core-test)

Testing ohmycards.web.core-test

Ran 1 tests containing 2 assertions.
0 failures, 0 errors.
nil[:test] Compiling ...
[:app] Compiling ...
[:app] Build completed. (477 files, 0 compiled, 0 warnings, 0.30s)
[:test] Build completed. (489 files, 2 compiled, 0 warnings, 0.82s)

cljs.user> 
cljs.user> 
cljs.user> (run-tests 'ohmycards.web.core-test)
------ REPL Error while processing ---------------------------------------------
(run-tests 'ohmycards.web.core-test)
Encountered error when macroexpanding cljs.test/test-ns-block.
AssertionError: Assert failed: Namespace ohmycards.web.core-test does not exist
(ana-api/find-ns ns)
	cljs.test/test-ns-block (test.cljc:373)
	cljs.test/test-ns-block (test.cljc:368)
	clojure.core/apply (core.clj:669)
	clojure.core/apply (core.clj:660)
	cljs.analyzer/macroexpand-1*/fn--2622 (analyzer.cljc:3919)
	cljs.analyzer/macroexpand-1* (analyzer.cljc:3917)
	cljs.analyzer/macroexpand-1* (analyzer.cljc:3904)
	cljs.analyzer/macroexpand-1 (analyzer.cljc:3968)
	cljs.analyzer/macroexpand-1 (analyzer.cljc:3964)
	cljs.analyzer/analyze-seq (analyzer.cljc:4001)
	cljs.analyzer/analyze-seq (analyzer.cljc:3981)
	cljs.analyzer/analyze-form (analyzer.cljc:4190)
	cljs.analyzer/analyze-form (analyzer.cljc:4187)
	cljs.analyzer/analyze* (analyzer.cljc:4243)
	cljs.analyzer/analyze* (analyzer.cljc:4235)
	cljs.analyzer/analyze (analyzer.cljc:4263)
	cljs.analyzer/analyze (analyzer.cljc:4246)
	cljs.analyzer/analyze (analyzer.cljc:4256)
	cljs.analyzer/analyze (analyzer.cljc:4246)
	cljs.analyzer/analyze (analyzer.cljc:4254)
	cljs.analyzer/analyze (analyzer.cljc:4246)
	shadow.build.cljs-hacks/shadow-parse-invoke*/fn--9482/fn--9483 (cljs_hacks.cljc:752)
	clojure.core/mapv/fn--8445 (core.clj:6912)
	clojure.core.protocols/fn--8159 (protocols.clj:168)
	clojure.core.protocols/fn--8159 (protocols.clj:124)
	clojure.core.protocols/fn--8114/G--8109--8123 (protocols.clj:19)
	clojure.core.protocols/seq-reduce (protocols.clj:31)
	clojure.core.protocols/fn--8144 (protocols.clj:75)
	clojure.core.protocols/fn--8144 (protocols.clj:75)
	clojure.core.protocols/fn--8088/G--8083--8101 (protocols.clj:13)
	clojure.core/reduce (core.clj:6828)
	clojure.core/mapv (core.clj:6903)
	clojure.core/mapv (core.clj:6903)
	shadow.build.cljs-hacks/shadow-parse-invoke*/fn--9482 (cljs_hacks.cljc:752)
	clojure.lang.Delay.deref (Delay.java:42)
	clojure.core/deref (core.clj:2320)
	clojure.core/deref (core.clj:2306)
	shadow.build.cljs-hacks/shadow-parse-invoke* (cljs_hacks.cljc:867)
	shadow.build.cljs-hacks/shadow-parse-invoke* (cljs_hacks.cljc:741)
	cljs.analyzer/parse-invoke (analyzer.cljc:3730)
	cljs.analyzer/parse-invoke (analyzer.cljc:3728)
	cljs.analyzer/analyze-seq* (analyzer.cljc:3975)
	cljs.analyzer/analyze-seq* (analyzer.cljc:3972)
	cljs.analyzer/analyze-seq*-wrap (analyzer.cljc:3979)
	cljs.analyzer/analyze-seq*-wrap (analyzer.cljc:3977)
	cljs.analyzer/analyze-seq (analyzer.cljc:4003)
	cljs.analyzer/analyze-seq (analyzer.cljc:3981)
	cljs.analyzer/analyze-form (analyzer.cljc:4190)
	cljs.analyzer/analyze-form (analyzer.cljc:4187)
	cljs.analyzer/analyze* (analyzer.cljc:4243)
	cljs.analyzer/analyze* (analyzer.cljc:4235)
	cljs.analyzer/analyze (analyzer.cljc:4263)
	cljs.analyzer/analyze (analyzer.cljc:4246)
	cljs.analyzer/analyze (analyzer.cljc:4256)
	cljs.analyzer/analyze (analyzer.cljc:4246)
	cljs.analyzer/analyze (analyzer.cljc:4254)
	cljs.analyzer/analyze (analyzer.cljc:4246)
	shadow.build.cljs-hacks/shadow-parse-invoke*/fn--9482/fn--9483 (cljs_hacks.cljc:752)
	clojure.core/mapv/fn--8445 (core.clj:6912)
	clojure.core.protocols/fn--8159 (protocols.clj:168)
	clojure.core.protocols/fn--8159 (protocols.clj:124)
	clojure.core.protocols/fn--8114/G--8109--8123 (protocols.clj:19)
	clojure.core.protocols/seq-reduce (protocols.clj:31)
	clojure.core.protocols/fn--8144 (protocols.clj:75)
	clojure.core.protocols/fn--8144 (protocols.clj:75)
	clojure.core.protocols/fn--8088/G--8083--8101 (protocols.clj:13)
	clojure.core/reduce (core.clj:6828)
	clojure.core/mapv (core.clj:6903)
	clojure.core/mapv (core.clj:6903)
	shadow.build.cljs-hacks/shadow-parse-invoke*/fn--9482 (cljs_hacks.cljc:752)
	clojure.lang.Delay.deref (Delay.java:42)
	clojure.core/deref (core.clj:2320)
	clojure.core/deref (core.clj:2306)
	shadow.build.cljs-hacks/shadow-parse-invoke* (cljs_hacks.cljc:867)
	shadow.build.cljs-hacks/shadow-parse-invoke* (cljs_hacks.cljc:741)
	cljs.analyzer/parse-invoke (analyzer.cljc:3730)
	cljs.analyzer/parse-invoke (analyzer.cljc:3728)
	cljs.analyzer/analyze-seq* (analyzer.cljc:3975)
	cljs.analyzer/analyze-seq* (analyzer.cljc:3972)
	cljs.analyzer/analyze-seq*-wrap (analyzer.cljc:3979)

nil
cljs.user> 
```

Am I doing something unsupported? Is this expected not to work?

Thanks! And thanks shadow-cljs ;)

__EDIT__  

Forgot to specify the versions:
```
;; Connected to nREPL server - nrepl://localhost:39283
;; CIDER 0.26.0snapshot (package: 20200619.928), nREPL 0.7.0
;; Clojure 1.10.1, Java 14.0.1
```

```
shadow-cljs - server version: 2.10.12 running at http://localhost:9630
```

**Comment by vitorqb** [comment]
I have done some investigation and bisecting the versions it seems the described flow works fine until version `2.8.94`, therefore [2.8.95](https://github.com/thheller/shadow-cljs/releases/tag/2.8.95) was the first version where the described flow started throwing errors.

**Comment by thheller** [comment]
I guess I'm kinda lost what your actual problem is. You have a `:app` and `:test` build yet you want to run tests in the `:app` REPL? How come you are not just using `:test` build? I'm assuming that is using `:target :browser-test`?

I changed the REPL init logic to not make `cljs.user` part of the build by default. It will only be compiled and loaded once the REPL connects and then adheres to REPL logic (ie. no hot-reload). If you want to make it part of your build you can "force" it via `:devtools {:preloads [cljs.user]}`. I generally recommend staying away from using `cljs.user` (or `user` in CLJ) as that is a "global" namespace which will affect all builds potentially causing problems later on.

It is unclear to me however why you have a `:test` build if don't use it?

**Comment by vitorqb** [comment]
@thheller Thanks for getting back to me.

> You have a :app and :test build yet you want to run tests in the :app REPL? 

Yes, I use the REPL connected to the `:app` build to run tests. It's quite handy because I can run only the tests I want to and I have the feedback instantly without leaving emacs. It also avoids running many other tests unrelated to what I'm working on. This way it saves me time and allows me to quickly find the failing tests. I like it quite better than having the tests in the browser and with notifications, and it fits nicely on my TDD-like workflow.

To achieve my workflow I imported all tests namespaces in the `cljs.user` (via a third ns). This way every time I changed a test file the `:app` build would recompile and I could re-run the test in the repl with `cljs.test`. 

I hope this clarifies my use case.

> I'm assuming that is using :target :browser-test?
> [...]
> It is unclear to me however why you have a :test build if don't use it?

My `:test` target is actually `:karma`, which I use when I want to run all tests in the application and be sure I did not break something. And it's specially there for CI as well. 

From your feedback, I will try two solutions when I have time:
1. Having a third build (something like `:repl-test`) that declares the test nss from it's entrypoint, and use a repl to this build to run the tests
2. Your solution with `:devtools {:preloads [cljs.user]}`
